### PR TITLE
add: added wave task and fix support for using the plugin in recursive models

### PR DESCRIPTION
**Pull request by kapeps**
Depends on:
- [ ] https://github.com/tidewise/simulation-gazebo_usv/pull/8
- [ ] https://github.com/rock-gazebo/simulation-orogen-rock_gazebo/pull/28
- [ ] https://github.com/rock-gazebo/simulation-rock_gazebo/pull/46

### CODE REVIEW

File: tasks/WaveSourceTask.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    wave_vel_msg.set_x(wave_velocity.x());
+    wave_vel_msg.set_y(wave_velocity.y());
+    wave_vel_msg.set_z(wave_velocity.z());
+    mWaveVelocityPublisher->Publish(wave_vel_msg);
+
+    gazebo::msgs::Vector3d wave_per_msg;
+    wave_per_msg.set_x(wave_frequency.x());
+    wave_per_msg.set_y(wave_frequency.y());
+    wave_per_msg.set_z(wave_frequency.z());
+    mWaveFrequencyPublisher->Publish(wave_per_msg);
```
Comment (jhonasiv) [inline tasks/WaveSourceTask.cpp:77]: I suppose the `per` is for period, does the message contain information about the wave period or frequency?

File: tasks/WaveSourceTask.hpp [hunk 2]
```diff
@@ -31,10 +30,11 @@ namespace gazebo_usv{
     protected:
         std::string mModelName;
         gazebo::transport::NodePtr mNode;
-        gazebo::transport::PublisherPtr mWaveVelocityPublisher;
+        gazebo::transport::PublisherPtr mWaveAmplitudePublisher;
```
Comment (afonsofonbraga) [inline tasks/WaveSourceTask.hpp:33]: According to the latest coding standard all variables must be snake_case and private/protected should start with m_

**Review by afonsofonbraga — CHANGES_REQUESTED** [review]

**Review by afonsofonbraga — APPROVED** [review]

**Review by afonsofonbraga — APPROVED** [review]

File: tasks/DirectForceApplicationTask.cpp [hunk 3]
```diff
@@ -39,8 +36,8 @@ bool DirectForceApplicationTask::configureHook()
         return false;
     }
 
-    auto topic_name = _link_name.get() + "/gazebo_usv_force";
-    mForcePub = mNode->Advertise<gazebo::msgs::Vector3d>("~/" + topic_name);
+    auto topic_name = mModelName + "/" + _link_name.get() + "/gazebo_usv_force";
```
[the remark this answers is not in the document]
Comment (kapeps) [reply tasks/DirectForceApplicationTask.cpp:39]: We are using it, as the mModelName is set at setGazeboModel

File: gazebo_usv.orogen [hunk 4]
```diff
@@ -19,13 +19,20 @@ task_context 'WindSourceTask', subclasses: 'rock_gazebo::ModelPluginTask' do
...
-    attribute 'name', '/std/string'
-
     # Wind velocity input port
     input_port 'wind_velocity', '/base/Vector3d'
 end
 
+task_context 'WaveSourceTask', subclasses: 'rock_gazebo::ModelPluginTask' do
+    needs_configuration
+
+    exception_states 'NO_TOPIC_CONNECTION'
```
[the remark this answers is not in the document]
Comment (kapeps) [reply gazebo_usv.orogen:29]: Some tasks dont use this exception_state, is it universal?

**Review by jhonasiv — APPROVED** [review]

**Review by seixasxbr — APPROVED** [review]

**Review by afonsofonbraga — APPROVED** [review]

**Pull request merged**